### PR TITLE
increase code coverage to 100%

### PR DESCRIPTION
**Pull request by iandrc**
Currently, 4 out of 4 files have 100% coverage:

| File       | % Stmts | % Branch | % Funcs | % Lines | Uncovered Line # |
|------------|---------|----------|---------|---------|------------------|
| All files  |   100 |    100 |     100 |   100 |                  |
| dynamic.js |   100 |    100 |     100 |   100 |               |
|   index.js |     100 |      100 |     100 |     100 |                  |
|  routes.js |     100 |      100 |     100 |     100 |                  |
|  static.js |     100 |      100 |     100 |     100 |                  |

Fixes: https://github.com/fastify/fastify-swagger/issues/258

#### Checklist

- [x] run `npm run test` and `npm run benchmark`
- [x] tests and/or benchmarks are included
- [x] commit message and code follows [Code of conduct](https://github.com/fastify/fastify/blob/master/CODE_OF_CONDUCT.md)

### CODE REVIEW

File: static.js [hunk 1]
```diff
@@ -65,6 +65,8 @@ module.exports = function (fastify, opts, next) {
     }
 
     fastify.register(require('./routes'), options)
+  } else {
+    console.log()
```
Comment (mcollina) [inline static.js:69]: why this empty `console.log` was added here?

File: test/static.js [hunk 2]
```diff
@@ -43,17 +43,88 @@ test('specification validation check works', t => {
...
+    exposeRoute: true,
+    mode: 'static',
+    specification: {
+      path: './examples/example-static-specification.yaml',
+      baseDir: 1
+    }
+  }
+
+  t.plan(1)
+  const fastify = new Fastify()
```
Comment (mcollina) [inline test/static.js:90]: remove `new`

File: test/static.js [hunk 3]
```diff
@@ -304,3 +381,242 @@ test('/documentation/:file should be served from custom location with trailing s
...
+    },
+    (err, res) => {
+      t.error(err)
+      t.is(typeof res.payload, 'string')
+      t.is(res.headers['content-type'], 'application/json; charset=utf-8')
+      try {
+        t.pass('valid swagger json')
+      } catch (err) {
+        t.fail(err)
+      }
```
Comment (mcollina) [inline test/static.js:478]: this try-catch is superflous

File: test/static.js [hunk 4]
```diff
@@ -304,3 +381,242 @@ test('/documentation/:file should be served from custom location with trailing s
...
```
Comment (mcollina) [inline test/static.js:461]: this try-catch is superflous

File: test/static.js [hunk 5]
```diff
@@ -304,3 +381,242 @@ test('/documentation/:file should be served from custom location with trailing s
...
+    (err, res) => {
+      t.error(err)
+      t.is(typeof res.payload, 'string')
+      t.is(res.headers['content-type'], 'application/x-yaml')
+      try {
+        yaml.safeLoad(res.payload)
+        t.pass('valid swagger yaml')
+      } catch (err) {
+        t.fail(err)
+      }
```
Comment (mcollina) [inline test/static.js:430]: this try-catch is superflous

File: test/static.js [hunk 6]
```diff
@@ -304,3 +381,242 @@ test('/documentation/:file should be served from custom location with trailing s
...
+    },
+    (err, res) => {
+      t.error(err)
+      t.is(typeof res.payload, 'string')
+      t.is(res.headers['content-type'], 'application/json; charset=utf-8')
+      try {
+        t.pass('valid swagger json')
+      } catch (err) {
+        t.fail(err)
+      }
```
Comment (mcollina) [inline test/static.js:557]: this try-catch is superflous

File: test/static.js [hunk 7]
```diff
@@ -304,3 +381,242 @@ test('/documentation/:file should be served from custom location with trailing s
...
```
Comment (mcollina) [inline test/static.js:574]: this try-catch is superflous

File: routes.js [hunk 8]
```diff
@@ -68,7 +68,7 @@ function fastifySwagger (fastify, opts, next) {
     schema: { hide: true },
     handler: function (req, reply) {
       const file = req.params['*']
-      if (file === '') {
+      if (file === 'null') {
```
Comment (mcollina) [inline routes.js:71]: why this change? I think this was correct before

File: test/static.js [hunk 9]
```diff
@@ -304,3 +381,242 @@ test('/documentation/:file should be served from custom location with trailing s
...
+    }
+  )
+})
+
+test('inserts default opts in fastifySwagger', t => {
+  t.plan(1)
+  const fastify = Fastify()
+  const next = () => console.log()
```
Comment (mcollina) [inline test/static.js:614]: why?

File: test/static.js [hunk 10]
```diff
@@ -304,3 +381,242 @@ test('/documentation/:file should be served from custom location with trailing s
...
+
+  // eslint-disable-next-line
+  fastify.register(Function(fastifySwagger(fastify, null, next)))
```
Comment (mcollina) [inline test/static.js:617]: Why the `Function`?

**Comment by mcollina** [comment]
> ** I have some issues with dynamic.js would be nice if you can help me out with it.**

what issues?

File: static.js (hunk 1 above)
Comment (iandrc) [reply static.js:69]: Otherwise, the coverage tool says that `else path not taken` and treats it as an uncovered branch. I didn't know how to do it the other way. That's why I inserted an empty `console.log`. Though, that's not the best option. 

I saw some tests in `fastify` repo, you don't have such a problem. Do you ignore such messages from the coverage somehow?

File: routes.js (hunk 8 above)
Comment (iandrc) [reply routes.js:71]: That's why I asked you to take a look. Actually, otherwise, I can't reach the code inside the `if` statement. I believe that's because it uses the default redirection rules defined above.

How do you suggest to change this line so it a) works correctly, b) does not harm the coverage?

File: test/static.js (hunk 9 above)
Comment (iandrc) [reply test/static.js:614]: I needed to pass `next` function as a param to `fastifySwagger`, and the `next` returns `void`. I didn't find a better solution to this.

File: test/static.js (hunk 10 above)
Comment (iandrc) [reply test/static.js:617]: If we don't use a `Function` constructor, `avvio` throws an error saying that:
```
 FAIL  test/static.js
✖ plugin must be a function or a promise
node_modules/avvio/boot.js                                                                       
  192 | function assertPlugin (plugin) {                                                           
  193 |   if (!(plugin && (typeof plugin === 'function' || typeof plugin.then === 'function'))) {  
> 194 |     throw new Error('plugin must be a function or a promise')                              
      | ----------^                                                                                
  195 |   }                                                                                        
  196 | }                                                                                          
  197 | 
```

So, if we remove the `Function` for this test, it'll fail

**Comment by iandrc** [comment]
> > ** I have some issues with dynamic.js would be nice if you can help me out with it.**
> 
> what issues?

Yeah, I've spent a couple of days trying to cover all the remaining lines/statements/branches inside `dynamic.js` file but I just can't reach them. So, it would be nice to have some new ideas from you or someone else how I can do it. 
Additionally, there are `8` uncovered branches (approx. half of them do not have an `else` block). What should I do with that half? Add some kind of `else { console.log() }` or we can ignore them?

**Comment by mcollina** [comment]
Which lines are unreacheable? You can paste links to the line in the issue.

---

> Additionally, there are 8 uncovered branches (approx. half of them do not have an else block). What should I do with that half? Add some kind of else { console.log() }

I don't understand why adding an `else` and a `console.log` helps. You need to add a test where the main condition in the `if` is not triggered.

**Comment by iandrc** [comment]
> Which lines are unreacheable? You can paste links to the line in the issue.
> 
> > Additionally, there are 8 uncovered branches (approx. half of them do not have an else block). What should I do with that half? Add some kind of else { console.log() }
> 
> I don't understand why adding an `else` and a `console.log` helps. You need to add a test where the main condition in the `if` is not triggered.

I'll try more then

File: routes.js (hunk 8 above)
Comment (mcollina) [reply routes.js:71]: I think you just need to send your http request to `/documentation` or `/documentation` so that the `*` param would be empty.

File: test/static.js (hunk 9 above)
Comment (mcollina) [reply test/static.js:590]: _never_ use `console.log`. You can add an assertion there with `t.error(err)` to assert that `next` is called without errors.

File: routes.js (hunk 8 above)
Comment (iandrc) [reply routes.js:71]: Nope, it should be smth like: `/documentation/a` at least in order to trigger those lines. Therefore, they won't be empty. 
That's why I've changed that check to 
```
if (file === 'null')
```

I tried different approaches but still couldn't cover those lines 'as is'. So, if you have any ideas just let me know

File: dynamic.js [hunk 11]
```diff
@@ -218,6 +218,7 @@ module.exports = function (fastify, opts, next) {
       }
 
       swaggerMethod.responses = genResponse(schema ? schema.response : null)
+      /* istanbul ignore next */
```
Comment (iandrc) [inline dynamic.js:221]: I've ignored the following check 'cause we'll always have a `swaggerRoute` in place, therefore, we have no need to take `else` path

File: dynamic.js [hunk 12]
```diff
@@ -281,14 +282,17 @@ module.exports = function (fastify, opts, next) {
         const rawJsonSchema = fastifyResponseJson[key]
         const resolved = ref.resolve(rawJsonSchema)
 
+        /* istanbul ignore next */
```
Comment (iandrc) [inline dynamic.js:285]: The following `if-else` is ignored as a `type` is somehow inserted under the hood. Therefore, we'll never fall into an `else` as far as I understood

File: dynamic.js [hunk 13]
```diff
@@ -281,14 +282,17 @@ module.exports = function (fastify, opts, next) {
...
         if (resolved.type || resolved.$ref) {
           responsesContainer[key] = {
             schema: resolved
           }
         } else {
+          /* istanbul ignore next */
           responsesContainer[key] = resolved
         }
 
+        /* istanbul ignore next */
```
Comment (iandrc) [inline dynamic.js:295]: We'll never have a `description` unless we directly specify it before the `if`

File: routes.js (hunk 8 above)
Comment (mcollina) [reply routes.js:71]: What is called when you call `/documentation` or `/documentation/` then?

Comment (iandrc) [reply routes.js:71]: Code on line `23` if I understand correctly:
```
fastify.route({
    url: '/',
    method: 'GET',
    schema: { hide: true },
    handler: (req, reply) => {
      reply.redirect(getRedirectPathForTheRootRoute(req.raw.url))
    }
  })
```

File: dynamic.js (hunk 11 above)
Comment (mcollina) [reply dynamic.js:221]: Then let's drop the if?

File: dynamic.js (hunk 12 above)
Comment (mcollina) [reply dynamic.js:285]: Let's drop the if.

File: dynamic.js (hunk 13 above)
Comment (mcollina) [reply dynamic.js:295]: let's drop the if.

File: routes.js (hunk 8 above)
Comment (mcollina) [reply routes.js:71]: Then we can drop the if.

File: dynamic.js (hunk 11 above)
Comment (iandrc) [reply dynamic.js:221]: I'll do so

File: dynamic.js (hunk 13 above)
Comment (iandrc) [reply dynamic.js:295]: Ok then

**Comment by mcollina** [comment]
how are we for code coverage now?

**Comment by iandrc** [comment]
Hey @mcollina 

We're approx. at the same point. I'm currently trying to cover the remaining statements and branches :)
Hopefully, I'll push them today or maybe tomorrow

File: dynamic.js [hunk 14]
```diff
@@ -106,6 +106,7 @@ module.exports = function (fastify, opts, next) {
       swaggerObject.externalDocs = externalDocs
     }
 
+    /* istanbul ignore next */
```
Comment (mcollina) [inline dynamic.js:109]: Why is this needed?

File: dynamic.js [hunk 15]
```diff
@@ -281,17 +281,11 @@ module.exports = function (fastify, opts, next) {
...
-        } else {
-          responsesContainer[key] = resolved
+        responsesContainer[key] = {
+          schema: resolved
         }
 
-        if (!responsesContainer[key].description) {
-          responsesContainer[key].description = 'Default Response'
-        }
+        responsesContainer[key].description = 'Default Response'
```
Comment (mcollina) [inline dynamic.js:288]: merge this with the object definition at line 285.

File: package.json [hunk 16]
```diff
@@ -5,7 +5,7 @@
   "main": "index.js",
   "scripts": {
     "prepare": "node prepare-swagger-ui",
-    "test": "standard && tap test/*.js && npm run typescript",
+    "test": "standard && tap --100 test/*.js && npm run typescript",
```
[the remark this answers is not in the document]
Comment (mcollina) [reply package.json:8]: Perfect!

File: dynamic.js (hunk 14 above)
[the remark this answers is not in the document]
Comment (mcollina) [reply dynamic.js:109]: You should add a test where `ref` is falsy.

[the remark this answers is not in the document]
Comment (iandrc) [reply dynamic.js:109]: @mcollina 
The `ref` is inside `dynamic.js` set on [line 13](https://github.com/fastify/fastify-swagger/blob/master/dynamic.js#L13) and has `undefined` value. 

I believe that was done on purpose. If we initialize it with some values, it may affect the existing logic. Therefore, we cannot change its value. 

So, what do you suggest?

[the remark this answers is not in the document]
Comment (mcollina) [reply dynamic.js:109]: I mean there should be a test where we are calling the `app.swagger()` function twice.

File: dynamic.js [hunk 17]
```diff
@@ -106,22 +106,20 @@ module.exports = function (fastify, opts, next) {
       swaggerObject.externalDocs = externalDocs
     }
 
-    if (!ref) {
```
Comment (iandrc) [inline dynamic.js:109]: I've played with code a bit more and found out that we don't need this `if` clause. The `swagger` function is called only once at the very beginning, and the second call does not allow us to trigger it once again. Therefore, we're quite safe to remove it, it should not break any logic.

**Comment by iandrc** [comment]
Hey @mcollina 

Should I add something or this can be merged to the main branch? Do I need to request some more reviews?

**Comment by mcollina** [comment]
I was waiting for more reviews.. but not really needed! Landed!
